capellambse/extensions/validation/_validate.py

```diff
@@ -0,0 +1,290 @@
+# SPDX-FileCopyrightText: Copyright DB Netz AG and the capellambse contributors
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+__all__ = [
+    "Category",
+    "ElementValidation",
+    "ModelValidation",
+    "register_rule",
+    "Result",
+    "Results",
+    "Rule",
+    "store_result",
+    "Type",
+    "Validation",
+    "Validator",
+]
+
+import abc
+import collections.abc as cabc
+import dataclasses
+import enum
+import typing as t
+
+from lxml import etree
+
+import capellambse
+from capellambse import helpers
+from capellambse.model import common as c
+
+
+class Category(enum.Flag):
+    """A category for a rule."""
+
+    REQUIRED = enum.auto()
+    RECOMMENDED = enum.auto()
+    SUGGESTED = enum.auto()
+
+
+@dataclasses.dataclass
+class Result:
+    """A validation rule result."""
+
+    uuid: helpers.UUIDString
+    category: t.Literal[Category.REQUIRED] | t.Literal[
+        Category.RECOMMENDED
+    ] | t.Literal[Category.SUGGESTED]
+    value: bool
+
+
+Validator = cabc.Callable[
+    [c.GenericElement], bool
+]  # | Callable [[Metric], Result]
+Type = t.Union[
```

Comment:
this Type union adds a lot of unnecessary complexity to the code. In rules, we always bind rules to real classes and not strings while in templates we always use strings. My suggestion is remove it and use the right "types" in the right contexts.